apps/SceneClustering/SceneClustering.cpp

```diff
@@ -0,0 +1,315 @@
+	OPEN_LOGFILE(MAKE_PATH(APPNAME _T("-")+Util::getUniqueName(0)+_T(".log")).c_str());
+
+	// print application details: version and command line
+	Util::LogBuild();
+	LOG(_T("Command line:%s"), Util::CommandLineToString(argc, argv).c_str());
+
+	// validate input
+	Util::ensureValidPath(OPT::strInputFileName);
+	Util::ensureUnifySlash(OPT::strInputFileName);
+	if (OPT::vm.count("help") || OPT::strInputFileName.IsEmpty()) {
+		boost::program_options::options_description visible("Available options");
+		visible.add(generic).add(config);
+		GET_LOG() << visible;
+	}
+
+	if (OPT::strInputFileName.IsEmpty())
+		return false;
+	
+	// initialize optional options
+	Util::ensureValidPath(OPT::strOutputDirectory);
+	Util::ensureUnifySlash(OPT::strOutputDirectory);
+	if (OPT::strOutputDirectory.IsEmpty())
+		OPT::strOutputDirectory = WORKING_FOLDER;
+
+	if(OPT::nMaxClusterSize < OPT::nMinClusterSize)
+	{
+		LOG("max-cluster-size value must be greater than min-cluster-size value");
+		return false;
+	}
+	// initialize global options
+	Process::setCurrentProcessPriority((Process::Priority)OPT::nProcessPriority);
+	#ifdef _USE_OPENMP
+	if (OPT::nMaxThreads != 0)
+		omp_set_num_threads(OPT::nMaxThreads);
+	#endif
+
+	#ifdef _USE_BREAKPAD
+	// start memory dumper
+	MiniDumper::Create(APPNAME, WORKING_FOLDER);
+	#endif
+
+	Util::Init();
+	return true;
+}
+
+// finalize application instance
+void Finalize()
+{
+	#if TD_VERBOSE != TD_VERBOSE_OFF
+	// print memory statistics
+	Util::LogMemoryInfo();
+	#endif
+
+	CLOSE_LOGFILE();
+	CLOSE_LOGCONSOLE();
+	CLOSE_LOG();
+}
+
+int main(int argc, LPCTSTR* argv)
+{
+	#ifdef _DEBUGINFO
+	// set _crtBreakAlloc index to stop in <dbgheap.c> at allocation
+	_CrtSetDbgFlag(_CRTDBG_ALLOC_MEM_DF | _CRTDBG_LEAK_CHECK_DF);// | _CRTDBG_CHECK_ALWAYS_DF);
+	#endif
+
+	if (!Initialize(argc, argv))
+		return EXIT_FAILURE;
+
+	Scene scene(OPT::nMaxThreads);
+	// load and texture the mesh
+	if (!scene.Load(MAKE_PATH_SAFE(OPT::strInputFileName)))
+		return EXIT_FAILURE;
+	
+	if (scene.pointcloud.IsEmpty()) {
+		LOG("error: empty initial point-cloud");
+		return EXIT_FAILURE;
+	}
+
+
+	const String baseFileName(MAKE_PATH_SAFE(Util::getFileFullName(OPT::strInputFileName)));
+
+	{
+	// compute clustering
+	TD_TIMER_START();
+
+	// set up data
+	std::vector<nomoko::Camera> domset_cameras; // Camera is useless in domset...
+	std::vector<nomoko::View> domset_views;
+	std::vector<nomoko::Point> domset_points; // nomoko::Point is an Eigen::Vec3f...
+	std::map<uint32_t, uint32_t> view_fwd_reindexing; // get continuous indexing
+	std::map<uint32_t, uint32_t> view_bkwd_reindexing; // reverse continuous indexing
+
+	uint32_t curr_ID = 0;
+	FOREACH(IdxC, scene.images)
+	{
+		const auto & curr_image = scene.images[IdxC];
+		if(curr_image.IsValid())
+		{
+			view_fwd_reindexing[IdxC] = curr_ID;
+			view_bkwd_reindexing[curr_ID] = IdxC; 
+			nomoko::View v;
+			v.rot = Eigen::Matrix<double,3,3,1>(curr_image.camera.R).cast<float>();
```

Comment:
I'd use REAL instead of double

Reply:
thanks! fixed in 9e84c2fcc0f16d451483c30359da754c5eab2286